File: tests/sdk/go/guardduty_findings_test.go
package sdk_test

import (
	"os"
	"testing"
)

func TestGuardDutyCreateInjectListGetFindings(t *testing.T) {
	requireReady(t)
	if os.Getenv("NOCTAXRIS_GUARDDUTY_INJECT") != "1" {
		t.Skip("set NOCTAXRIS_GUARDDUTY_INJECT=1 on the API process for lab InjectFindings")
	}

	createStatus, createBody, createParsed := signedJSONTarget(t, "guardduty", "GuardDuty.CreateDetector", map[string]any{})
	if createStatus != 200 {
		t.Fatalf("CreateDetector status=%d body=%s", createStatus, createBody)
	}
	detectorID, _ := createParsed["DetectorId"].(string)
	if detectorID == "" {
		t.Fatalf("CreateDetector missing DetectorId: %v", createParsed)
	}

	injStatus, injBody, injParsed := signedJSONTarget(t, "guardduty", "NoctaxrisGuardDuty.InjectFindings", map[string]any{
		"DetectorId": detectorID,
		"Findings": []map[string]any{{
			"type":     "UnauthorizedAccess:IAMUser/InstanceCredentialExfiltration.InsideAWS",
			"severity": 8,
			"title":    "sdk cred exfil",
		}},
	})
	if injStatus != 200 {
		t.Fatalf("InjectFindings status=%d body=%s", injStatus, injBody)
	}
	injIDs, _ := injParsed["FindingIds"].([]any)
	if len(injIDs) != 1 {
		t.Fatalf("InjectFindings FindingIds=%v", injParsed)
	}
	findingID, _ := injIDs[0].(string)
	if findingID == "" {
		t.Fatalf("InjectFindings empty FindingId: %v", injParsed)
	}

	listStatus, listBody, listParsed := signedJSONTarget(t, "guardduty", "GuardDuty.ListFindings", map[string]any{
		"DetectorId": detectorID,
	})
	if listStatus != 200 {
		t.Fatalf("ListFindings status=%d body=%s", listStatus, listBody)
	}
	listed, _ := listParsed["FindingIds"].([]any)
	listedOK := false
	for _, id := range listed {
		if s, ok := id.(string); ok && s == findingID {
			listedOK = true
			break
		}
	}
	if !listedOK {
		t.Fatalf("ListFindings missing %s: %v", findingID, listParsed)
	}

	getStatus, getBody, getParsed := signedJSONTarget(t, "guardduty", "GuardDuty.GetFindings", map[string]any{
		"DetectorId": detectorID,
		"FindingIds": []string{findingID},
	})
	if getStatus != 200 {
		t.Fatalf("GetFindings status=%d body=%s", getStatus, getBody)
	}
	findings, _ := getParsed["Findings"].([]any)
	if len(findings) != 1 {
		t.Fatalf("GetFindings Findings=%v", getParsed)
	}
	f, _ := findings[0].(map[string]any)
	ftype, _ := f["type"].(string)
	if ftype == "" {
		t.Fatalf("GetFindings missing type: %v", f)
	}
}

File: tests/sdk/nodejs/test/guardduty-findings.test.mjs
import { test } from "node:test";
import assert from "node:assert/strict";
import { requireReady, signedJsonTarget } from "../lib/helpers.mjs";

test("GuardDuty CreateDetector InjectFindings List GetFindings", async (t) => {
  if (!(await requireReady(t))) return;
  if (process.env.NOCTAXRIS_GUARDDUTY_INJECT !== "1") {
    t.skip(
      "set NOCTAXRIS_GUARDDUTY_INJECT=1 on the API process for lab InjectFindings",
    );
    return;
  }

  const created = await signedJsonTarget(
    "guardduty",
    "GuardDuty.CreateDetector",
    {},
  );
  assert.equal(
    created.status,
    200,
    `CreateDetector status=${created.status} body=${created.body}`,
  );
  const detectorId = created.json?.DetectorId;
  assert.ok(detectorId, `CreateDetector missing DetectorId: ${created.body}`);

  const injected = await signedJsonTarget(
    "guardduty",
    "NoctaxrisGuardDuty.InjectFindings",
    {
      DetectorId: detectorId,
      Findings: [
        {
          type: "UnauthorizedAccess:IAMUser/InstanceCredentialExfiltration.InsideAWS",
          severity: 8,
          title: "sdk cred exfil",
        },
      ],
    },
  );
  assert.equal(
    injected.status,
    200,
    `InjectFindings status=${injected.status} body=${injected.body}`,
  );
  const findingIds = injected.json?.FindingIds || [];
  assert.equal(findingIds.length, 1, `FindingIds=${JSON.stringify(findingIds)}`);
  const findingId = findingIds[0];
  assert.ok(findingId);

  const listed = await signedJsonTarget(
    "guardduty",
    "GuardDuty.ListFindings",
    { DetectorId: detectorId },
  );
  assert.equal(
    listed.status,
    200,
    `ListFindings status=${listed.status} body=${listed.body}`,
  );
  assert.ok(
    (listed.json?.FindingIds || []).includes(findingId),
    `ListFindings missing ${findingId}: ${listed.body}`,
  );

  const got = await signedJsonTarget("guardduty", "GuardDuty.GetFindings", {
    DetectorId: detectorId,
    FindingIds: [findingId],
  });
  assert.equal(
    got.status,
    200,
    `GetFindings status=${got.status} body=${got.body}`,
  );
  const findings = got.json?.Findings || [];
  assert.equal(findings.length, 1);
  assert.ok(findings[0]?.type, `GetFindings missing type: ${got.body}`);
});
